mithril-aggregator/src/http_server/routes/artifact_routes/cardano_database.rs

```diff
@@ -0,0 +1,367 @@
+    warp::path!("artifact" / "cardano-database" / String)
+        .and(warp::get())
+        .and(middlewares::with_logger(dependency_manager))
+        .and(middlewares::with_http_message_service(dependency_manager))
+        .and(middlewares::with_metrics_service(dependency_manager))
+        .and_then(handlers::get_artifact_by_signed_entity_id)
+}
+
+fn serve_cardano_database_dir(
+    router_state: &RouterState,
+) -> impl Filter<Extract = (impl warp::Reply,), Error = warp::Rejection> + Clone {
+    warp::path("cardano_database_download")
+        .and(warp::fs::dir(
+            router_state.configuration.snapshot_directory.clone(),
+        ))
+        .and(middlewares::with_logger(router_state))
+        .and(middlewares::extract_config(router_state, |config| {
+            config.allow_http_serve_directory
+        }))
+        .and_then(handlers::ensure_downloaded_file_is_a_cardano_database)
+}
+
+mod handlers {
+    use crate::http_server::routes::reply;
+    use crate::services::MessageService;
+    use crate::MetricsService;
+    use slog::{debug, warn, Logger};
+    use std::convert::Infallible;
+    use std::sync::Arc;
+    use warp::http::StatusCode;
+
+    pub const LIST_MAX_ITEMS: usize = 20;
+
+    /// List artifacts
+    pub async fn list_artifacts(
+        logger: Logger,
+        http_message_service: Arc<dyn MessageService>,
+    ) -> Result<impl warp::Reply, Infallible> {
+        match http_message_service
+            .get_cardano_database_list_message(LIST_MAX_ITEMS)
+            .await
+        {
+            Ok(message) => Ok(reply::json(&message, StatusCode::OK)),
+            Err(err) => {
+                warn!(logger,"list_artifacts_cardano_database"; "error" => ?err);
+                Ok(reply::server_error(err))
+            }
+        }
+    }
+
+    /// Get artifact by signed entity id
+    pub async fn get_artifact_by_signed_entity_id(
+        signed_entity_id: String,
+        logger: Logger,
+        http_message_service: Arc<dyn MessageService>,
+        metrics_service: Arc<MetricsService>,
+    ) -> Result<impl warp::Reply, Infallible> {
+        metrics_service
+            .get_artifact_detail_cardano_db_total_served_since_startup()
+            .increment();
+
+        match http_message_service
+            .get_cardano_database_message(&signed_entity_id)
+            .await
+        {
+            Ok(Some(signed_entity)) => Ok(reply::json(&signed_entity, StatusCode::OK)),
+            Ok(None) => {
+                warn!(logger, "cardano_database_details::not_found");
+                Ok(reply::empty(StatusCode::NOT_FOUND))
+            }
+            Err(err) => {
+                warn!(logger,"cardano_database_details::error"; "error" => ?err);
+                Ok(reply::server_error(err))
+            }
+        }
+    }
+
+    /// Download a file if it's a Cardano_database archive file
+    pub async fn ensure_downloaded_file_is_a_cardano_database(
```

Comment:
Could the function be unit tested?

Reply:
I guess that it is tested in the e2e test (same as for the snapshots).
As far as the structure is not complete yet, I'd rather wait until this is the case: I have added a TODO in that sense :+1: